xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -615,7 +633,102 @@ unsigned int CGUIFontTTF::GetTextureLineHeight() const
   return m_cellHeight + spacing_between_characters_in_texture;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr)
+std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecText& text)
+{
+  std::vector<Glyph> glyphs;
+  if (text.empty())
+  {
+    return glyphs;
+  }
+  std::vector<hb_script_t> scripts;
+  std::vector<RunInfo> runs;
+  hb_unicode_funcs_t* ufuncs = hb_unicode_funcs_create(hb_unicode_funcs_get_default());
```

Comment:
Should this be a global singleton? Seems a bit wasteful to allocate it each time we render *every* character.

Reply:
I don’t think the `hb_unicode_funcs_create()` is necessary, calling `hb_unicode_funcs_get_default()` should be enough here.

Reply:
> Should this be a global singleton? Seems a bit wasteful to allocate it each time we render _every_ character.

Is global singleton necessary as this is allocated not for every character but for every _text_ 




> I don’t think the `hb_unicode_funcs_create()` is necessary, calling `hb_unicode_funcs_get_default()` should be enough here.

Ok. I will update the commit

Reply:
> I don’t think the `hb_unicode_funcs_create()` is necessary, calling `hb_unicode_funcs_get_default()` should be enough here.

@khaledhosny removing `hb_unicode_funcs_create()` causes runtime error
```
../harfbuzz/src/hb-object.hh:262: Type* hb_object_reference(Type*) [with Type = hb_unicode_funcs_t]: Assertion `hb_object_is_valid (obj)' failed.
Aborted (core dumped)
```
Any idea?

Reply:
never mind. figured it out

> > I don’t think the `hb_unicode_funcs_create()` is necessary, calling `hb_unicode_funcs_get_default()` should be enough here.
> 
> @khaledhosny removing `hb_unicode_funcs_create()` causes runtime error
> 
> ```
> ../harfbuzz/src/hb-object.hh:262: Type* hb_object_reference(Type*) [with Type = hb_unicode_funcs_t]: Assertion `hb_object_is_valid (obj)' failed.
> Aborted (core dumped)
> ```
> 
> Any idea?

Reply:
Remove the `hb_unicode_funcs_destroy(ufuncs);`, since you are using the default functions manged by HarfBuzz itself.